### PR TITLE
WIP: Change how CARRY4 modelling is handled

### LINKED ISSUE #1597: Carry Chain COUT drives multiple net sinks
When updating the baslitex test with the newest version of Litex, I have encountered the following assertion failure in VPR:

```
vpr/src/place/place_macro.cpp:135 find_all_the_macro: Assertion 'cluster_ctx.clb_nlist.net_sinks(curr_net_id).size() == 1' failed.
```

The assertion checks whether nets between macros (in this case carry chains) have more than one sink. This because a net connecting to blocks in a macro should not drive other sinks, as this would create inconsistencies with the macro placement.

I have tracked down the possible cause of the issue being in the fact that COUT pins from carry chain get to be connected to external nets rather than just the CIN pin of a next carry in the chain (if no other carry is present in the chain, the COUT pin should not be connected to any other sink).

eblif:
```
.subckt CARRY4_VPR CO_CHAIN=$abc$108049$auto$alumacc.cc:485:replace_alu$19819.slice[0].carry4_1st_full.COUT CO_FABRIC0=$techmap451788$auto$alumacc.cc:485:replace_alu$19819.slice[0].carry4_1st_full.CO[0] CO_FABRIC1=$techmap451788$auto$alumacc.cc:485:replace_alu$19819.slice[0].carry4_1st_full.CO[1] CO_FABRIC2=$techmap451788$auto$alumacc.cc:485:replace_alu$19819.slice[0].carry4_1st_full.CO[2] CO_FABRIC3=$techmap451788$auto$alumacc.cc:485:replace_alu$19819.slice[0].carry4_1st_full.CO[3] DI0=a7ddrphy_bitslip8_value[0] DI1=$false DI2=$false DI3=$false O0=$abc$108049$auto$alumacc.cc:485:replace_alu$19819.slice[0].carry4_1st_full.O[0] O1=$abc$108049$auto$alumacc.cc:485:replace_alu$19819.slice[0].carry4_1st_full.O[1] O2=$abc$108049$auto$alumacc.cc:485:replace_alu$19819.slice[0].carry4_1st_full.O[2] O3=$abc$108049$auto$alumacc.cc:485:replace_alu$19819.slice[0].carry4_1st_full.O[3] S0=$abc$108049$auto$alumacc.cc:485:replace_alu$19819.slice[0].carry4_1st_full.S[0] S1=a7ddrphy_bitslip8_value[1] S2=a7ddrphy_bitslip8_value[2] S3=a7ddrphy_bitslip8_value[3]

...

.names $abc$108049$__35631__ $abc$108049$auto$alumacc.cc:485:replace_alu$19819.slice[0].carry4_1st_full.COUT a7ddrphy_bitslip8_r[13] a7ddrphy_bitslip8_r[21] $abc$108049$__39304__
```

From the eblif, we can see that the COUT pin is present in multiple nets, so this issue most probably relates to the yosys step.

top_synth.v:
```
  CARRY4_VPR #(
    .CYINIT_AX(1'h0),
    .CYINIT_C0(1'h1),
    .CYINIT_C1(1'h0)
  ) _34817_ (
    .CO_CHAIN(_09496_),
    .CO_FABRIC0(_14535_[0]),
    .CO_FABRIC1(_14535_[1]),
    .CO_FABRIC2(_14535_[2]),
    .CO_FABRIC3(_14535_[3]),
    .DI0(a7ddrphy_bitslip7_value[0]),
    .DI1(1'h0),
    .DI2(1'h0),
    .DI3(1'h0),
    .O0(_09497_[0]),
    .O1(_09497_[1]),
    .O2(_09497_[2]),
    .O3(_09497_[3]),
    .S0(_09498_[0]),
    .S1(a7ddrphy_bitslip7_value[1]),
    .S2(a7ddrphy_bitslip7_value[2]),
    .S3(a7ddrphy_bitslip7_value[3])
  );

...

assign _04983_ = 32'd288577023 >> { _03833_, _09496_, a7ddrphy_bitslip7_r[15], a7ddrphy_bitslip7_r[19], a7ddrphy_bitslip7_r[21] };
```

Also in the verilog representation, the CO_CHAIN output is connected to an assign statement.

### PR DESCRIPTION
**Pull request by litghost**
This PR attempts to fix #1597 by change the strategy used to decongest SLICEL CARRY4 outputs.  Previously, all ALU CO uses were modeled as a `O ^ S`, but yosys's abc9 pass is now good enough to undo this change.  So rather than continue to try to fight abc9 as it gets better at merging logic, this PR uses a new CARRY4 congestion avoidance strategy.

Changes:
 - Yosys arith mapping using upstream CARRY4 mapping
 - New pass (implemented as a Python3 script) de-congests O + CO carry rows by using a LUT equation from the inputs
 - CARRY4_VPR is modelled as a blackbox instead of a whitebox to prevent abc9 from modifying/swapping outputs during cleanup optimization

### CODE REVIEW

**Comment by litghost** [comment]
There is a bug, working on it.

**Comment by litghost** [comment]
I still need to add a test that the CO LUT explosion calculates correctly.

File: xc/xc7/techmap/fix_carry.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
+import json
```
Comment (acomodi) [inline xc/xc7/techmap/fix_carry.py:1]: Docstrings missing on several functions

File: xc/xc7/yosys/synth.tcl [hunk 2]
```diff
@@ -33,7 +33,38 @@ if { [info exists ::env(INPUT_XDC_FILE)] && $::env(INPUT_XDC_FILE) != "" } {
...
+techmap -map  $::env(TECHMAP_PATH)/carry_map.v
+write_json $::env(OUT_JSON).carry_fixup.json
+exec $::env(PYTHON3) $::env(TECHMAP_PATH)/fix_carry.py < $::env(OUT_JSON).carry_fixup.json > $::env(OUT_JSON).carry_fixup_out.json
+design -push
+read_json $::env(OUT_JSON).carry_fixup_out.json
+
+techmap -map  $::env(TECHMAP_PATH)/clean_carry_map.v
+
+# Re-read baseline libraries
+read_verilog -lib -specify -D_EXPLICIT_CARRY +/xilinx/cells_sim.v
```
Comment (acomodi) [inline xc/xc7/yosys/synth.tcl:48]: Why is EXPLICIT_CARRY needed here if we are transitioning to have the native CARRY map from yosys?

File: xc/xc7/yosys/synth.tcl [hunk 3]
```diff
@@ -33,7 +33,38 @@ if { [info exists ::env(INPUT_XDC_FILE)] && $::env(INPUT_XDC_FILE) != "" } {
...
+}
+
+# Re-run optimization flow to absorb carry modifications
+hierarchy -check -auto-top
+
+write_ilang $::env(OUT_JSON).pre_abc9.ilang
+if { $::env(USE_ROI) == "TRUE" } {
+    synth_xilinx -flatten -abc9 -nosrl -noclkbuf -nodsp -noiopad -nowidelut -run map_ffs:check
+} else {
+    synth_xilinx -flatten -abc9 -nosrl -noclkbuf -nodsp -iopad -nowidelut -run map_ffs:check
```
Comment (acomodi) [inline xc/xc7/yosys/synth.tcl:62]: How does this re-synthesis step impact on run-time?

File: xc/xc7/yosys/synth.tcl [hunk 4]
```diff
@@ -33,7 +33,38 @@ if { [info exists ::env(INPUT_XDC_FILE)] && $::env(INPUT_XDC_FILE) != "" } {
...
+}
+
+write_ilang $::env(OUT_JSON).post_abc9.ilang
+
+chtype -map CARRY4_VPR CARRY4_FIX
```
Comment (acomodi) [inline xc/xc7/yosys/synth.tcl:67]: Why the cell type needs to be changed to CARRY4_FIX here?

File: xc/xc7/techmap/cells_sim.v [hunk 5]
```diff
@@ -1911,3 +1979,12 @@ module PS7_VPR (
 
 endmodule
 
+module BANK();
```
Comment (acomodi) [inline xc/xc7/techmap/cells_sim.v:1982]: What is this module related to and why is it needed?

File: xc/xc7/techmap/iobs.v [hunk 6]
```diff
@@ -1,4 +1,5 @@
 module IBUF (
+  (* iopad_external_pin *)
```
Comment (acomodi) [inline xc/xc7/techmap/iobs.v:2]: Are these changes to IOBs intended? They shouldn't be related to the CARRY_CHAIN changes.

I also see a failure in CI related to IOBUFs:
```
ERROR: Module `IBUF' referenced in module `top' in cell `$iopadmap$top.cpu_reset' does not have a parameter named 'IO_LOC_PAIRS'.
```

Could the error be related to these changes?

File: xc/xc7/yosys/synth.tcl (hunk 2 above)
Comment (litghost) [reply xc/xc7/yosys/synth.tcl:48]: It's a mistake

File: xc/xc7/yosys/synth.tcl (hunk 3 above)
Comment (litghost) [reply xc/xc7/yosys/synth.tcl:62]: I haven't measured it yet.

File: xc/xc7/yosys/synth.tcl (hunk 4 above)
Comment (litghost) [reply xc/xc7/yosys/synth.tcl:67]: I'll add a comment

File: xc/xc7/techmap/cells_sim.v (hunk 5 above)
Comment (litghost) [reply xc/xc7/techmap/cells_sim.v:1982]: The `check` after re-reading the JSON notices that the "read_xdc" plugin added the "BANK" module, which does exists, and yosys complains that it has parameters that do not exist.

File: xc/xc7/techmap/iobs.v (hunk 6 above)
Comment (litghost) [reply xc/xc7/techmap/iobs.v:2]: These changes are required to avoid problems with the iopadmap double inserting.  @mkurc-ant made similiar changes in his IDDR/ODDR PR I believe.  These fixes are unrelated to the carry chain stuff, except in that they are needed bug fixes for other reasons.

Comment (litghost) [reply xc/xc7/techmap/iobs.v:2]: I understand the issue with the `IO_LOC_PAIRS`, working on a fix.

File: xc/xc7/yosys/synth.tcl (hunk 2 above)
Comment (litghost) [reply xc/xc7/yosys/synth.tcl]: Fixed

**Comment by acomodi** [comment]
@litghost I have pushed an updated version of litex with the fixes to the clock which now is correctly set to 60MHz as well as the changes in the mem.init files. The commit is this one: https://github.com/SymbiFlow/symbiflow-arch-defs/pull/1605/commits/aa2fb936db257db07aecc57c81d6df5d478dfff1

**Comment by litghost** [comment]
> @litghost I have pushed an updated version of litex with the fixes to the clock which now is correctly set to 60MHz as well as the changes in the mem.init files. The commit is this one: [aa2fb93](https://github.com/SymbiFlow/symbiflow-arch-defs/commit/aa2fb936db257db07aecc57c81d6df5d478dfff1)

Great, I'll ingest those changes.

File: xc/xc7/techmap/fix_carry.py (hunk 1 above)
Comment (litghost) [reply xc/xc7/techmap/fix_carry.py:25]: Yep, fixed now

**Comment by litghost** [comment]
I believe all review comments have been addressed other than the re-abc9 cost.

**Comment by litghost** [comment]
> One other thing we would also need to verify is the QoR (as well as runtime) reached with these changes. I believe we can just inspect the qor.csv file generated as an artifact from CI.

Agreed.  Once it's green I'll post the results.

**Comment by litghost** [comment]
The new install failure is now fixed.  The diff fasm failure is understood, but no solution is immediately obvious.  More concerning is that the place and route time for designs goes up with this change.  Investigating potentially solutions.

**Comment by litghost** [comment]
A fasm2bels bug showed up in the CI, fix is here: https://github.com/SymbiFlow/symbiflow-xc-fasm2bels/pull/15

**Comment by litghost** [comment]
Superceeded by #1674
